common/flatpak-transaction.c

```diff
@@ -3024,6 +3031,125 @@ flatpak_transaction_add_install_flatpakref (FlatpakTransaction *self,
+       * user's control and we no longer install it again, even if the user
+       * manually removes it. */
+      if (!priv->reinstall &&
+          flatpak_g_ptr_array_contains_string (preinstalled_refs, flatpak_decomposed_get_ref (decomposed)))
+        {
+          g_info ("Preinstall ref %s is marked as already preinstalled; skipping", flatpak_decomposed_get_ref (decomposed));
+          continue;
+        }
+
+      for (int j = 0; remotes[j] != NULL; j++)
+        {
+          const char *remote = remotes[j];
+          g_autoptr(GError) local_error = NULL;
+          g_autofree char *remote_collection_id = NULL;
+
+          if (flatpak_dir_get_remote_disabled (priv->dir, remote))
+            continue;
+
+          remote_collection_id = flatpak_dir_get_remote_collection_id (priv->dir, remote);
+
+          /* Choose the first match if the collection ID was not specified */
+          if (collection_id != NULL &&
+              g_strcmp0 (remote_collection_id, collection_id) != 0)
+            continue;
+
+          g_info ("Adding preinstall of %s from remote %s", flatpak_decomposed_get_ref (decomposed), remote);
+
+          if (!flatpak_transaction_add_ref (self, remote, decomposed, NULL, NULL, NULL,
+                                            FLATPAK_TRANSACTION_OPERATION_INSTALL,
+                                            NULL, NULL, FALSE, TRUE, NULL,
+                                            &local_error))
+            g_info ("Failed to add preinstall ref %s: %s", flatpak_decomposed_get_ref (decomposed),
+                     local_error->message);
+        }
+    }
+
+  /* Find previously preinstalled refs that are no longer in the preinstall
+   * list and should now get uninstalled */
+  for (guint i = 0; i < preinstalled_refs->len; i++)
+    {
+      const char *ref = g_ptr_array_index (preinstalled_refs, i);
+
+      /* No longer in the preinstall.d list, so uninstall */
+      if (!flatpak_g_ptr_array_contains_string (install_refs, ref))
+        {
+          g_autoptr(GError) local_error = NULL;
+          g_autoptr(FlatpakDecomposed) decomposed = NULL;
+
+          decomposed = flatpak_decomposed_new_from_ref (ref, error);
+          if (decomposed == NULL)
+            return FALSE;
+
+          g_info ("Preinstalled ref %s is no longer listed as wanted in preinstall.d config; uninstalling", flatpak_decomposed_get_ref (decomposed));
+
+          if (!flatpak_transaction_add_ref (self, NULL, decomposed, NULL, NULL, NULL, FLATPAK_TRANSACTION_OPERATION_UNINSTALL, NULL, NULL, FALSE, TRUE, NULL, &local_error))
+            {
+              if (g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_NOT_INSTALLED))
+                {
+                  g_clear_error (&local_error);
+                }
+              else
+                {
+                  g_propagate_error (error, g_steal_pointer (&local_error));
+                  return FALSE;
```

Comment:
More of a general question but shouldn't this try to do as much as possible, even if some things report an error?

Reply:
Yes, doing as much as possible is exactly what this code is trying to achieve - two lines above this is `if (g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_NOT_INSTALLED)) g_clear_error (&local_error); ` which makes it continue for the simple error case where the ref has already gotten uninstalled. If there's any other kind of error (the `else` error path) I think we should error out - in that case it's probably something majorly wrong with the flatpak installation and it's worth erroring out and letting the user know.

Reply:
Okay, so there are a number of errors which you consider bad enough to abort completely because it might indicate something wrong with the installation. That makes sense.